lib/viewmodels/challenge_view_model.dart

```diff
@@ -1,36 +1,90 @@
+import "package:geoflutterfire_plus/geoflutterfire_plus.dart";
 import "package:hooks_riverpod/hooks_riverpod.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
 import "package:proxima/models/ui/challenge_card_data.dart";
+import "package:proxima/services/database/challenge_repository_service.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+import "package:proxima/services/geolocation_service.dart";
+import "package:proxima/viewmodels/login_view_model.dart";
 
-// TODO: For now, this is duplicated code with the mock from the tests.
-// This will just be replaced when we implement a real view model anyway.
+/// This viewmodel is used to fetch the list of challenges that are displayed in
+/// the challenge feed. It fetches the challenges from the database and sorts
+/// them by putting the finished challenges at the end of the list. It transforms
+/// the challenges into [ChallengeCardData] objects to be displayed, by getting
+/// the posts from the post repository and calculating the distances as well as
+/// remaining time.
 class ChallengeViewModel extends AsyncNotifier<List<ChallengeCardData>> {
   @override
   Future<List<ChallengeCardData>> build() async {
-    return const [
-      ChallengeCardData.solo(
-        title: "I hate UNIL, here's why",
-        distance: 700,
-        timeLeft: 27,
-        reward: 250,
-      ),
-      ChallengeCardData.solo(
-        title:
-            "John fell here lmaoooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo",
-        distance: 3200,
-        timeLeft: 27,
-        reward: 400,
-      ),
-      ChallengeCardData.group(
-        title: "I'm moving out",
-        distance: 4000,
-        reward: 350,
-      ),
-      ChallengeCardData.soloFinished(
-        title: "What a view!",
-        timeLeft: 27,
-        reward: 200,
-      ),
-    ];
+    final geoLocationService = ref.watch(geoLocationServiceProvider);
+    final challengeRepository = ref.watch(challengeRepositoryServiceProvider);
+
+    final currentPosition = await geoLocationService.getCurrentPosition();
+    final currentUser = ref.watch(uidProvider);
+
+    final firestoreChallenges = await challengeRepository.getChallenges(
+      currentUser!,
+      currentPosition,
+    );
+
+    final postRepository = ref.watch(postRepositoryProvider);
+    final now = DateTime.now();
+    Iterable<Future<ChallengeCardData>> futureUiChallenges =
+        firestoreChallenges.map((challenge) async {
```

Comment:
This line is very long. Not specifying its type put using a final instead could make it more readable in my opinion:
```suggestion
    final futureUiChallenges = firestoreChallenges.map((challenge) async {
```